Add aggregated search filter event and user counts to pings

attfarhan opened the pull request:
- Adds counts of number of searches conducted using each filter, aggregated daily, weekly, and monthly, to pings
- Adds counts of number of users conducting searches containing each filter, aggregated daily, weekly, and monthly.

attfarhan commented on cmd/frontend/types/types.go:
```diff
@@ -221,17 +221,34 @@ type SearchUsageStatistics struct {
+	Structural         *SearchEventStatistics
+	Symbol             *SearchEventStatistics
+	Timeout            *SearchEventStatistics
+	Type               *SearchEventStatistics
```
@efritz I know in the last PR you mentioned I should use more specific types than `SearchEventStatistics` if we're not measuring latency. I wanted to just use `SearchEventCounts` for the new filters that weren't already added here. However, when creating the map of filter fields to handler functions, I wasn't sure how to specify multiple possible return types, since some filters (e.g. `repo`, `file`), already exist in `SearchUsagePeriods` to measure latency and are of type `SearchEventStatistics`.  Any recommendations here? Or if you know the best way to handle this in Go?

efritz replied:
Could you create two loops with different map types?

efritz commented on cmd/frontend/db/event_logs.go:
```diff
@@ -547,3 +547,42 @@ func (l *eventLogs) ListUniqueUsersAll(ctx context.Context, startDate, endDate t
 	return users, nil
 }
+
+// CountUniqueUsersSearchingWithFilterPerPeriod counts the number of unique users who have searched using a each search filter,
```
Search-specific items shouldn't belong in this package. This should be in the search.go file in the usage stats package.

efritz commented on cmd/frontend/types/types.go:
```diff
@@ -221,17 +221,34 @@ type SearchUsageStatistics struct {
-	Commit      *SearchEventStatistics
-	Symbol      *SearchEventStatistics
-	SearchModes *SearchModeUsageStatistics
+	StartTime          time.Time
```
Note: you'll also need to add ALL of these to the bigquery schema before merging.

attfarhan replied:
Yep, got it

attfarhan commented:
@efritz addressed your comments. will add to the BigQuery schema PR once this is approved.

In addition to your comments, I also removed `ByEventNameWithArgument` and replaced it with `ByEventNameWithCondition`, which is more general, but still requires using in conjunction with `ByEventName`.

efritz commented on cmd/frontend/db/event_logs.go:
```diff
@@ -268,12 +268,12 @@ func (l *eventLogs) CountUniqueUsersPerPeriod(ctx context.Context, periodType Pe
+			if opt.EventFilters.ByEventNameWithCondition != nil {
 				if opt.EventFilters.ByEventName == "" {
-					return nil, fmt.Errorf("The ByEventNameWithArgument option must be used together with the ByEventName option.")
+					return nil, fmt.Errorf("The ByEventNameWithCondition option must be used together with the ByEventName option.")
```
why enforce this?

attfarhan replied:
you're right no need to anymore, thanks.

efritz commented on cmd/frontend/db/event_logs.go:
```diff
@@ -547,3 +547,42 @@ func (l *eventLogs) ListUniqueUsersAll(ctx context.Context, startDate, endDate t
 	return users, nil
 }
+
+// // CountUniqueUsersSearchingWithFilterPerPeriod counts the number of unique users who have searched using a each search filter,
```
lol commenting out the docstring doesn't really address it 😝

efritz replied:
I can be overridden on this, but I specifically tried to make changes to these files agnostic to their use rather than tied to a specific subset of events (code intel, campaigns, search). If it's too difficult to put into the consumer I get that.

attfarhan replied:
Oops I meant to delete these. They’re unused.

rvantonder commented:
Looks like `patterntype` is not in this? Is that intentional @attfarhan?

attfarhan commented:
Not intentional, thanks @rvantonder

attfarhan commented:
Merging this into the parent PR
